xlators/storage/posix/src/posix-aio.c

```diff
@@ -402,6 +406,138 @@ posix_aio_writev(call_frame_t *frame, xlator_t *this, fd_t *fd,
+        0,
+    };
+    int _fd = -1;
+    int op_ret = -1;
+    int op_errno = 0;
+    int ret = 0;
+    fd_t *fd = NULL;
+
+    GF_ASSERT(paiocb);
+
+    frame = paiocb->frame;
+    this = frame->this;
+    prebuf = paiocb->prebuf;
+    fd = paiocb->fd;
+    _fd = paiocb->_fd;
+
+    if (res < 0) {
+        op_ret = -1;
+        op_errno = -res;
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, P_MSG_FSYNC_FAILED,
+               "%s(async) failed fd=%d (%d)",
+               (paiocb->iocb.aio_lio_opcode == IO_CMD_FDSYNC ? "fdatasync"
+                                                             : "fsync"),
+               _fd, res);
+        goto out;
+    }
+
+    ret = posix_fdstat(this, fd->inode, _fd, &postbuf);
+    if (ret != 0) {
+        op_ret = -1;
+        op_errno = errno;
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, P_MSG_FSTAT_FAILED,
+               "fstat failed on fd=%d", _fd);
+        goto out;
+    }
+
+    op_ret = res;
+    op_errno = 0;
+
+out:
+    STACK_UNWIND_STRICT(fsync, frame, op_ret, op_errno, &prebuf, &postbuf,
+                        NULL);
+
+    if (paiocb->fd)
+        fd_unref(paiocb->fd);
+
+    GF_FREE(paiocb);
+
+    return 0;
+}
+
+int
+posix_aio_fsync(call_frame_t *frame, xlator_t *this, fd_t *fd, int32_t datasync,
+                dict_t *xdata)
+{
+    int32_t op_errno = EINVAL;
+    int _fd = -1, ret = -1;
+    struct posix_fd *pfd = NULL;
+    struct posix_aio_cb *paiocb = NULL;
+    struct posix_private *priv = NULL;
+    struct iocb *iocb = NULL;
+
+    VALIDATE_OR_GOTO(frame, err);
+    VALIDATE_OR_GOTO(this, err);
+    VALIDATE_OR_GOTO(fd, err);
+
+    priv = this->private;
+
+    ret = posix_fd_ctx_get(fd, this, &pfd, &op_errno);
+    if (ret < 0) {
+        gf_msg(this->name, GF_LOG_WARNING, op_errno, P_MSG_PFD_NULL,
+               "pfd is NULL from fd=%p", fd);
+        goto err;
+    }
+
+    _fd = pfd->fd;
+
+    paiocb = posix_aio_cb_init(frame, fd, _fd, GF_FOP_FSYNC);
+    if (!paiocb) {
+        op_errno = ENOMEM;
+        goto err;
+    }
+
+    paiocb->iocb.aio_lio_opcode = datasync ? IO_CMD_FDSYNC : IO_CMD_FSYNC;
+
+    iocb = &paiocb->iocb;
+
+    ret = posix_fdstat(this, fd->inode, _fd, &paiocb->prebuf);
+    if (ret != 0) {
+        op_errno = errno;
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, P_MSG_FSTAT_FAILED,
+               "fstat failed on fd=%p", fd);
+        goto err;
+    }
+
+    LOCK(&fd->lock);
+    {
+        ret = io_submit(priv->ctxp, 1, &iocb);
+    }
+    UNLOCK(&fd->lock);
+
+    if (ret != 1) {
+        op_errno = -ret;
+        gf_msg(this->name, GF_LOG_ERROR, op_errno, P_MSG_IO_SUBMIT_FAILED,
+               "io_submit() returned %d,gfid=%s", ret,
+               uuid_utoa(fd->inode->gfid));
+        goto err;
+    }
+
+    return 0;
+
+err:
+    STACK_UNWIND_STRICT(fsync, frame, -1, op_errno, 0, 0, 0);
+
+    if (paiocb) {
+        if (paiocb->fd)
+            fd_unref(paiocb->fd);
+        GF_FREE(paiocb);
+    }
```

Comment:
It would be good to have a corresponding posix_aio_cb_fini() and move these lines there and from posix_aio_$FOP_complete() functions, for all FOPS.

Reply:
OK